### DISCUSSION TITLE
Why not use gunicorn/workers with multiple containers?

### DISCUSSION THREAD

**Discussion by shawnwall** [fastapi/fastapi#9760] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# this is not a question that has a specific code example, as it is related to deployment
```

### Description

* Dockerize fastapi application with gunicorn server workers, based on available vcpu with `multiprocessing`
* Deploy via AWS ECS/Fargate
* Deploy multiple containers based on autoscaling
* I expect this to be a supported path, to utilize each container's CPU/RAM effectively
* ...but your documentation states "you probably would want to have a single (Uvicorn) process per container, as you would already be handling replication at the cluster level." but you didn't quite state why its a bad idea to have multiple containers with multiple processes via https://fastapi.tiangolo.com/deployment/docker/#one-process-per-container and https://fastapi.tiangolo.com/deployment/docker/#containers-with-multiple-processes-and-special-cases
* I'm not saying you are wrong at all, but I'm just trying to figure out why it would be a bad idea to have multiple containers running via something like AWS ECS, each running multiple processes with gunicorn+uvicorn?

### Operating System

Linux

### FastAPI Version

0.95

### Python Version

3.9

**Comment by jgould22** [comment, marked as answer]
It is not really a "bad" idea. I just is not optimal.

You are essentially layering two different "scaling" technologies on top of each other and adding complexity for no benefit. 

If you are using something like k8s or a similar cloud offering using single workers allows you finer grain control. If you are using gunicorn with say 4 workers your minimum incremental unit to scale up is 4 workers as opposed to 1 worker if you have a single uvicorn instance. 

There is also the added complexity that Gunicorn adds. Things like the prometheus client (for metrics) work best with a single worker.

**Reply by shawnwall** [reply]
thank you for the reply. so overall, do you feel that running _more_ lower powered uvicorn based instances is preferred than less higher powered gunicorn/worker based? for more context, i'm in a scenario where some of my endpoints are slow due to synchronous communication with external services (db's, opensearch, etc) so more workers gives me more flex room per instance for db connection pools, etc. I guess I could increase the pool size on single thread uvicorn instances though...

**Reply by jgould22** [reply]
So I use Gunicorn when I am running services on my own hardware/vm.

I do this because I want to use the full power of the machine/vm that my service is deployed to since I have already paid the "cost" to provision that compute. 

In a cloud environment it is common to have the cloud provider provision compute for you on demand. This means if requests are queuing up the cloud provider would start more instances of the single uvicorn to handle this load. You want it to be able to do this in small increments because you do not want to over pay for compute you are not using in the cloud.   

When you say "slow due to synchronous communication" are you doing this synchronous communication inside a thread pool (ie with your controller is defined with a def) or an in the main loop?

**Reply by shawnwall** [reply]
Correct, my controllers are currently just `def` so they end up going through the thread pool executor built into fastapi.

The issue we've had with scaling is that CPU stays low; it ends up being our db connections, etc that cause issues. So we're going to have to likely do some pool size tweaking, along with utilizing other factors for scaling purposes.

**Reply by jgould22** [reply]
Ah yeah.

It is a bit finicky adjusting the size of that thread pool.

You may find this useful https://github.com/encode/starlette/issues/1724#issuecomment-1179063924

**Comment by ghomem** [comment]
I am involved in a similar project (FastAPI, ECS with Fargate) and the same question popped. I am not fully convinced about the answer above, for the same reasons that @shawnwall mentioned.

Let's say we have X containers and from time to time Y long running network-bound API clients contact those containers:

- if X >= Y everything is fine
- if X < Y there will be (Y-X) API clients that receive an error because all the X single process containers are busy (even if they are busy with a network load and their CPU is fully available)

In the situation above it would be beneficial to have multiple workers per container because any API client that would connect for a short operation would be attended. Of course, if we have N workers per container we would still saturate if NxY API long running API clients start using the API - but the problem here is that having NxY containers is expensive, while having processes inside a container (that might be busy but still have CPU available) does not add to that cost.

You will say: but you can autoscale... well yes, but the autoscaling is not trivial and requires application awareness to avoid running jobs being killed upon scale in ( https://github.com/aws/containers-roadmap/issues/125). Thus, we can't use autoscale for sudden load changes without applicational safeguards.

**Reply by shawnwall** [reply]
my guess here is people will say 'run n lower power/memory containers' than 'run x higher power/memory containers' to try to balance out cost

**Reply by ghomem** [reply]
OK, so let's compare three scenarios with the same cost:

1. 12 containers with 0.5 vcpu and a single worker
2. 2 containers with 3 vcpu and a single worker
3. 2 containers with 3 vcpu and 100 gunicorn workers each

What happens:

1. blocks with 12 network intensive API clients even if the CPU is mostly free, delivers bad single-threaded performance on a CPU bound client
2. blocks with 2 network intensive API clients even if the CPU is mostly free, delivers better single-threaded performance than scenario 1 on a CPU intensive client
3. blocks with 600 network bound API clients (or if the network saturates first), delivers better single-threaded performance than scenario 1 on a CPU intensive client, shares the CPU between concurrent CPU-intensive clients until the 6 vcpus are saturated

So, for a generic API that has network-intensive and cpu-intensive endpoints is there anything better than having 2 large containers with many gunicorn workers inside? (2 instead of one brings redundancy, of course)

I am open to being told that I am missing  something or not reasoning correctly, of course :-)

**Comment by ghomem** [comment]
Apart from the topic of number of workers, I am also not sure if this part:

_Deploy multiple containers based on autoscaling_

is generic enough to be a recommendation.

I am saying this because autoscaling is slow: **scale out** responds slowly to increasing demand and meanwhile many requests can be lost (especially if the containers are single process...), **scale in** needs specific application protections to avoid busy containers being killed if the average CPU consumption calculated over all the containers goes below the defined threshold (https://github.com/aws/containers-roadmap/issues/125). Autoscaling does not work for bursty demand with long [*] running processes - all the containers will be taken before there is an opportunity to scale out. It works well for predictable slow variations over a day but I am not sure if that would be the general case as to transform it in a recommendation.

[*] the definition of long depends on the context
